reloadData / forceReload on data providers

signedav opened the pull request:
This is especially useful as part of the API when the data are updated in the background (data is entered to the database elsewhere) and QGIS needs to be reloaded (including feature count in the legend).

All calls of `forceReload()` or `reloadData()` target `QgsDataProvider::reloadData()` what calls the implemented `reloadProviderData()` of the provider and `dataChanged()` signal is called always. That this signal is called changes the behavior of the `dataReload()` calls of all providers.

Mostly the featureCount reset is implemented in the `reloadProviderData()` instead of do it separately.

signedav committed:
forceRelaod function including requesting feature reset

Saijin-Naib commented:
Any chance this could be adapted to other database providers like GeoPackage?
https://github.com/qgis/QGIS/issues/29616  
https://github.com/qgis/QGIS/issues/29617  
https://github.com/qgis/QGIS/issues/29629

signedav committed:
Deprecate `forceReload()` and merge it together with `reloadData()`
`reloadData()` implementations are set to private on `reloadProviderData()`

All calls of `forceReload()` or `reloadData()` target `QgsDataProvider::reloadData()` what calls the implemented `reloadProviderData()` of the provider and `dataChanged()` signal is called allways. That this signal is called changes the behavior of the `dataReload()` calls of all providers.

signedav committed:
sip deprecated of forceReload and some doc improvements

signedav commented:
@Saijin-Naib Partly. If you replace the file in the background silently nothing happens. But on reloading with F5  it used to redraw the canvas before but did not had any effects on the feature count. Now - with this changes - it resets the featurecount and on toggeling a reload of the legendtree (e.g. the mentioned setting on/off of featureCount, but already clicking on it) displays the correct value.

m-kuhn commented:
Just to add to that, the other thing would be a new feature request. Similar to live layers / reload on postgres notify to have a file system watcher trigger a data reload. Feel free to open a new feature request issue for this (if there is not already one).

m-kuhn commented on python/core/auto_generated/qgsdataprovider.sip.in:
```diff
@@ -265,8 +265,11 @@ databases and servers.
-synchronize with changes in the data source
+Reloads the data from the source by calling reloadProviderData() implemented
+by providers with data caches to synchronize, changes in the data source, feature
+counts and other specific actions and it calls the dataChanged signal
```
```suggestion
counts and other specific actions and it.
Emits the `dataChanged` signal
```

To make it clearer that this is done by this method and not by subclasses.

signedav replied:
Should we use markdown `` for method names in the docs?

m-kuhn replied:
I think so.
Or even better something like "Emits the QgsDataProvider::dataChanged() signal" to autolink
http://www.doxygen.nl/manual/autolink.html

m-kuhn commented on src/core/qgsdataprovider.h:
```diff
@@ -614,6 +616,10 @@ class CORE_EXPORT QgsDataProvider : public QObject
     mutable QMutex mOptionsMutex;
 
+    /**
+     * Reloads the data according to the provider
```
```suggestion
     * Reloads the data according to the provider
     * \since QGIS 3.12
```

m-kuhn commented on src/core/qgsvectordataprovider.h:
```diff
@@ -554,8 +554,9 @@ class CORE_EXPORT QgsVectorDataProvider : public QgsDataProvider, public QgsFeat
      * This can be required if the underlying file is replaced.
-     */
-    virtual void forceReload();
+     * \deprecated QGIS 3.12 - will be removed in QGIS 4.0 - use reloadData instead
```
I would suggest to remove all the rest of the docs above to have one canonical source for information on reloadData

m-kuhn commented on src/providers/arcgisrest/qgsafsprovider.h:
```diff
@@ -89,6 +88,11 @@ class QgsAfsProvider : public QgsVectorDataProvider
+
+    /**
+     * Clears cache
+    */
```
FYI, when overriding virtual methods, they will inherit the docs from the parent. Adding specific docs will overwrite the parent docs. Therefore, only add docstrings if they have more information than the generic one in the parent.

signedav replied:
Thanks.
In this case it's fine I guess since the providers make different things to reload the data.

signedav committed:
improve docs

signedav committed:
avoiding double call of dataChanged

signedav committed:
removing the featureCount() request since it would recalculate it instead of resetting the count so it could be recalculated on the next featureCount()

signedav commented on src/providers/wfs/qgswfsprovider.cpp:
```diff
@@ -755,11 +757,8 @@ bool QgsWFSProvider::setSubsetString( const QString &theSQL, bool updateFeatureC
     QgsMessageLog::logMessage( errorMsg, tr( "WFS" ) );
-  reloadData();
-  if ( updateFeatureCount )
-    featureCount();
```
Hi @rouault 

While planing to integrate the featureCount "reset" in the `reloadProviderData` function, I noticed a behavior that is not consistent. While we have usually at this position a functionality resetting the feature counter, so it would be recalculated on the next call of `featureCount()` it recalculates it here. Means, it makes the opposite (a `featureCount()` call would not need to calculate again). 

To make it consistent with the other providers it would be needed to make a flag telling `featureCount` to recalculate or to remove this `featureCount()` call here and let it recalculate always on this provider on a `featureCount` call. I decided for the second one not to implement something that might be wrong. Do you approve that?

signedav commented:
No stale. 

@rouault any capacities to review that?

m-kuhn reviewed (approved):
This looks much cleaner than the former approach.
I'd say merge it on Wednesday if there are no objections.
